dv/formal/conductor.py

```diff
@@ -0,0 +1,802 @@
+        buckets = {}
+        done = []
+        for chunks, name in props:
+            if len(chunks) == 1:
+                done.append(name)
+            elif chunks[0] not in buckets:
+                buckets[chunks[0]] = [(chunks[1:], name)]
+            else:
+                buckets[chunks[0]].append((chunks[1:], name))
+        for pre in buckets:
+            bucket = group(buckets[pre])
+            # while len(bucket) == 1:
+            #     bucket = bucket[0]
+            done.append(bucket)
+        return done
+
+    chunked_names = [(chunk_name(name), name) for name in names]
+    return group(chunked_names)
+
+# ---------------------------------- Main ------------------------------
+
+parser = argparse.ArgumentParser(prog="conductor.py", description="Constructs and executes proofs.")
+parser.add_argument("mode", choices=["prove", "explore", "bmc", "info", "logopt"],
+                    help="Proof mode. "
+                    "prove will run existing proofs where they exist, "
+                    "explore will attempt to discover new proofs, "
+                    "bmc will run BMC on each property individually, "
+                    "info dumps stats about cached proofs, "
+                    "logopt extracts information (e.g. the k for k-induction) to help optimise proofs.")
+parser.add_argument("--fresh", action="store_true", help="In explore mode, do not use already constructed proofs, always construct new proofs.")
+parser.add_argument("--no-store", action="store_true", help="In explore mode, do not store constructed proof strategies.")
+parser.add_argument("--by-step", action="store_true", help="In prove mode, do proofs one step at a time")
+parser.add_argument("-p", "--properties", nargs="*", default=[], help="Restrict to only the properties with the given names, otherwise all properties. Especially helpful for BMC.")
+parser.add_argument("--missing", action="store_true", help="Equivalent to -p <each property that is not skipped but has no proof in a step where there are some proofs>")
+parser.add_argument("-s", "--start", type=int, default=0, help="First step to start at. (default: 0)")
+parser.add_argument("--bmc-step", type=int, default=1, help="Step size for BMC. (default: 1)")
+parser.add_argument("--bmc-start", type=int, default=4, help="Start length for BMC. (default: 4)")
+parser.add_argument("--hard", action="store_true", help="In explore mode, try harder to prove properties (1hr timeout, more engines).")
+parser.add_argument("--no-run", action="store_true", help="In explore mode don't run proofs again to check steps.")
+parser.add_argument("--no-kill", action="store_true", help="Don't kill proof processes due to running out of memory.")
+parser.add_argument("--check-complete", action="store_true", help="In prove mode, fail if there are unskipped properties without proofs.")
+parser.add_argument("--max-mem", type=float, default=0, help="Max memory in GB, otherwise use all available memory")
```

Comment:
You already have a maximum memory, I guess the only thing necessary is to link this up to the Github action.